examples/access-plugin-minimal/main.go

```diff
@@ -0,0 +1,274 @@
+			resp.HTTPStatusCode,
+		)
+	}
+
+	return nil
+
+}
+
+func (g *googleSheetsPlugin) updateRow(ar types.AccessRequest, rowNum int64) error {
+	row := g.makeRowData(ar)
+
+	req := sheets.BatchUpdateSpreadsheetRequest{
+		Requests: []*sheets.Request{
+			{
+				UpdateCells: &sheets.UpdateCellsRequest{
+
+					Fields: "*",
+					Start: &sheets.GridCoordinate{
+						RowIndex: rowNum,
+					},
+					Rows: []*sheets.RowData{
+						row,
+					},
+				},
+			},
+		},
+	}
+
+	resp, err := g.sheetsClient.BatchUpdate(spreadSheetID, &req).Do()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if resp.HTTPStatusCode == 201 || resp.HTTPStatusCode == 200 {
+		fmt.Println("Successfully updated a row")
+	} else {
+		fmt.Printf(
+			"Unexpected response code updating a row: %v\n",
+			resp.HTTPStatusCode,
+		)
+	}
+
+	return nil
+
+}
+
+func (g *googleSheetsPlugin) updateSpreadsheet(ar types.AccessRequest) error {
+	s, err := g.sheetsClient.Get(spreadSheetID).IncludeGridData(true).Do()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(s.Sheets) != 1 {
+		return trace.Wrap(
+			errors.New("the spreadsheet must have a single sheet"),
+		)
+	}
+
+	for _, d := range s.Sheets[0].Data {
+		for i, r := range d.RowData {
+			if r.Values[0] != nil &&
+				r.Values[0].UserEnteredValue != nil &&
+				r.Values[0].UserEnteredValue.StringValue != nil &&
+				*r.Values[0].UserEnteredValue.StringValue == ar.GetName() {
+				if err := g.updateRow(ar, int64(i)); err != nil {
+					return trace.Wrap(err)
+				}
+				fmt.Println("Updated a spreadsheet row.")
+			}
+		}
+	}
+	return nil
+}
+
+func (g *googleSheetsPlugin) handleEvent(ctx context.Context, event types.Event) error {
+
+	if event.Resource == nil {
+		return nil
+	}
+
+	r := event.Resource.(types.AccessRequest)
+
+	if r.GetState() == types.RequestState_PENDING {
+		return g.createRow(r)
+	}
+
+	return g.updateSpreadsheet(r)
+}
+
+func (g *googleSheetsPlugin) run() error {
+	ctx := context.Background()
+	proc := lib.NewProcess(ctx)
+	watcherJob := watcherjob.NewJob(
+		g.teleportClient,
+		watcherjob.Config{
+			Watch: types.Watch{Kinds: []types.WatchKind{types.WatchKind{Kind: types.KindAccessRequest}}},
+		},
+		g.handleEvent,
+	)
+
+	proc.SpawnCriticalJob(watcherJob)
+
+	fmt.Println("Started the watcher job")
+
+	<-watcherJob.Done()
+
+	fmt.Println("The watcher job is finished")
+
+	return nil
+}
+
+func main() {
+	ctx := context.Background()
+	svc, err := sheets.NewService(ctx, option.WithCredentialsFile("credentials.json"))
+	if err != nil {
+		panic(err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, initTimeout)
+	defer cancel()
+
+	creds := client.LoadIdentityFile("auth.pem")
+
+	teleport, err := client.New(ctx, client.Config{
+		Addrs:       []string{proxyAddr},
+		Credentials: []client.Credentials{creds},
+		DialOpts: []grpc.DialOption{
+			grpc.WithReturnConnectionError(),
+		},
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	gs := googleSheetsPlugin{
```

Comment:
nit: From the perspective of using this example as a base for one's own custom plugin, it would be cool to make the code a bit more modular. For example, rather than having a `googleSheetsPlugin`, we could have something like:

```
type accessRequestPlugin struct {
	teleportClient *client.Client
        eventHandler interface {
                HandleEvent(ctx context.Context, event types.Event) error
        }
}

type googleSheetsClient struct {
        sheetsClient   *sheets.SpreadsheetsService
}

func (*googleSheetsClient) HandleEvent(ctx context.Context, event types.Event) error {
       ...
}

plugin := accessRequestPlugin{
       teleportClient: ...,
       eventHandler: googleSheetsClient{
              sheetsClient: ...,
       }
}
```

Then the user could just swap out `eventHandler` with their own implementation very intuitively. This would also make splitting the code into separate files more natural.

I think this PR is a big improvement regardless, so this is more of a follow up suggestion than a blocker.

Reply:
This sounds great, especially if we want to make it the recommended way for a user to build/test their own Access Request plugin. If this is the way we want to go, I can look into making this change. 

When writing this guide, though, I had imagined the example as something that would demonstrate some key libraries in a runnable way, but not the basis for an actual production plugin. Is there anything else we'd need to add to the `accessRequestPlugin` type to make it more generally usable?

Reply:
Nothing off the top of my head, just having a teleport Client and some customizable event handler seems sufficient as a base.